### PR TITLE
Gracefully handle failure of IMDSv2 and allow fallback to IMDSv1

### LINKED ISSUE #560: 3.6.4 does not fetch iam credentials using IMDSv2 when running from inside containers with IMDSv2 Defaults
```
xxxxe.net failed | msg: non-zero return code | stderr: [fog][WARNING] Unable to fetch credentials: read timeout reached
rake aborted!
ArgumentError: Missing required arguments: aws_access_key_id, aws_secret_access_key
/usr/local/bundle/ruby/2.6.0/gems/fog-core-2.2.0/lib/fog/core/service.rb:244:in `validate_options'
/usr/local/bundle/ruby/2.6.0/gems/fog-core-2.2.0/lib/fog/core/service.rb:268:in `handle_settings'
/usr/local/bundle/ruby/2.6.0/gems/fog-core-2.2.0/lib/fog/core/service.rb:98:in `new'
/usr/local/bundle/ruby/2.6.0/gems/fog-core-2.2.0/lib/fog/core/services_mixin.rb:29:in `new'
/usr/local/bundle/ruby/2.6.0/gems/asset_sync-2.11.0/lib/asset_sync/storage.rb:20:in `connection'
```

3.6.3 works fine with instance credentials, however 3.6.4 does not.

### PR DESCRIPTION
**Pull request by atyndall**
Fixes #560 

To address edge cases where the IMDSv2 API does not work for users, I have added support for the `CredentialFetcher` to fall-back to not using IMDSv2 tokens when the API call to retrieve the token fails 3 times, or when the option `:disable_imds_v2` is set to true. This is similar to the approach employed by [`aws-sdk-core`](https://github.com/aws/aws-sdk-ruby/blob/master/gems/aws-sdk-core/lib/aws-sdk-core/instance_profile_credentials.rb).

### CODE REVIEW

**Comment by atyndall** [comment]
@v-kumar can you verify if this PR fixes your issue if your hop limit is still set to 1?

**Comment by geemus** [comment]
@atyndall looks good to me, thanks for this.

@v-kumar I'll wait for your confirmation for now to ensure we have fixed the issues sufficiently.

File: lib/fog/aws/credential_fetcher.rb [hunk 1]
```diff
@@ -58,6 +56,23 @@ def fetch_credentials(options)
...
+
+        def fetch_credentials_token_header(connection, disable_imds_v2)
+          return nil if disable_imds_v2
+
+          token = connection.put(
+            :path => INSTANCE_METADATA_TOKEN,
+            :idempotent => true,
+            :expects => 200,
+            :retry_interval => 1,
+            :retry_limit => 3,
```
Comment (v-kumar) [inline lib/fog/aws/credential_fetcher.rb:68]: Should you add a `read_timeout` here? Excon has a default read time out of 60 seconds and this is too long a time to wait to verify if the first put option works.

Comment (atyndall) [reply lib/fog/aws/credential_fetcher.rb:68]: I've added more timeouts now

**Comment by geemus** [comment]
Thanks!

**Comment by v-kumar** [comment]
Hi: 

Thanks for fixing the timeout. That actually was necessary.

When I tried it without the time out, it was a long time before it retried and by then other safety mechanisms kicked in and aborted the S3 sync process.

**Comment by geemus** [comment]
@v-kumar Is everything now working as expected for you then? I could cut a release, but was hoping to confirm it was in a good place before we did so.

**Comment by v-kumar** [comment]
I confirmed this worked. I just made sure I tested all four variations - 3.6.3, 3.6.4, the previous commit in PR, latest master.

**1. 3.6.3 - of course works.**

**2. 3.6.4 - Fails with ArgumentError: Missing required arguments**
```
ArgumentError: Missing required arguments: aws_access_key_id, aws_secret_access_key
/usr/local/bundle/ruby/2.6.0/gems/fog-core-2.2.0/lib/fog/core/service.rb:244:in `validate_options'
/usr/local/bundle/ruby/2.6.0/gems/fog-core-2.2.0/lib/fog/core/service.rb:268:in `handle_settings'
/usr/local/bundle/ruby/2.6.0/gems/fog-core-2.2.0/lib/fog/core/service.rb:98:in `new'
```

**3. PR first commit - without read_timeout/connect_timeout - Fails with timeout error in Excon**
```
Caused by:
IO::EAGAINWaitReadable: Resource temporarily unavailable - read would block
/usr/local/bundle/ruby/2.6.0/gems/excon-0.73.0/lib/excon/socket.rb:64:in `readline'
/usr/local/bundle/ruby/2.6.0/gems/excon-0.73.0/lib/excon/response.rb:63:in `parse'
/usr/local/bundle/ruby/2.6.0/gems/excon-0.73.0/lib/excon/middlewares/response_parser.rb:7:i
```

**4. Latest Master - Works**

**Comment by v-kumar** [comment]
Thanks for patiently working through the issues. @atyndall  and @geemus

**Comment by geemus** [comment]
Great, thanks for confirming and working through this.

**Comment by geemus** [comment]
released in 3.6.5

**Pull request merged**